src/token-processor/util/metadata-helpers.ts

```diff
@@ -0,0 +1,452 @@
+  trait_type: Type.String(),
+  value: Type.Any(),
+  display_type: Type.Optional(Type.String()),
+});
+const RawMetadataAttributes = Type.Array(RawMetadataAttribute);
+const RawMetadataAttributesCType = TypeCompiler.Compile(RawMetadataAttributes);
+
+// Raw metadata property types.
+const RawMetadataProperties = Type.Record(Type.String(), Type.Any());
+const RawMetadataPropertiesCType = TypeCompiler.Compile(RawMetadataProperties);
+
+type RawMetadataLocale = {
+  metadata: RawMetadataType;
+  locale?: string;
+  default: boolean;
+  uri: string;
+};
+
+/**
+ * Fetches all the localized metadata JSONs for a token. First, it downloads the default metadata
+ * JSON and parses it looking for other localizations. If those are found, each of them is then
+ * downloaded, parsed, and returned for DB insertion.
+ * @param uri - token metadata URI
+ * @param token - token DB entry
+ * @returns parsed metadata ready for insertion
+ */
+export async function fetchAllMetadataLocalesFromBaseUri(
+  uri: string,
+  token: DbToken
+): Promise<DbMetadataLocaleInsertBundle[]> {
+  const tokenUri = getTokenSpecificUri(uri, token.token_number);
+  const rawMetadataLocales: RawMetadataLocale[] = [];
+
+  const defaultMetadata = await getMetadataFromUri(tokenUri);
+  rawMetadataLocales.push({
+    metadata: defaultMetadata,
+    default: true,
+    uri: tokenUri,
+  });
+
+  // Does it declare localizations? If so, fetch and parse all of them.
+  if (RawMetadataLocalizationCType.Check(defaultMetadata.localization)) {
+    const uri = defaultMetadata.localization.uri;
+    const locales = defaultMetadata.localization.locales;
+    rawMetadataLocales[0].locale = defaultMetadata.localization.default;
+    for (const locale of locales) {
+      if (locale === rawMetadataLocales[0].locale) {
+        // Skip the default, we already have it.
+        continue;
+      }
+      const localeUri = getTokenSpecificUri(uri, token.token_number, locale);
+      const localeMetadata = await getMetadataFromUri(localeUri);
+      rawMetadataLocales.push({
+        metadata: localeMetadata,
+        locale: locale,
+        default: false,
+        uri: localeUri,
+      });
+    }
+  }
+
+  return parseMetadataForInsertion(rawMetadataLocales, token);
+}
+
+/**
+ * Returns a metadata URI that is specific to a token number within a contract,
+ * i.e. replacing `{id}` with the token number and `{locale}` with the given
+ * locale.
+ * @param uri - Original metadata URI
+ * @param tokenNumber - token number
+ * @param locale - locale to apply
+ * @returns token specific uri string
+ */
+export function getTokenSpecificUri(uri: string, tokenNumber: number, locale?: string): string {
+  const tokenNumStr = tokenNumber.toString();
+  const localeStr = locale ?? '';
+  return (
+    uri
+      .replaceAll('{id}', tokenNumStr)
+      .replaceAll('{locale}', localeStr)
+      // Patterns below are not SIP-016 compliant.
+      .replaceAll('$TOKEN_ID', tokenNumStr)
```

Comment:
Not sure how strict we're wanting to be here, but could use regex for case-insensitive:

```suggestion
    uri
      .replaceAll(/{id}/ig, tokenNumStr)
      .replaceAll(/{locale}/ig, localeStr)
      // Patterns below are not SIP-016 compliant.
      .replaceAll(/\$TOKEN_ID/ig, tokenNumStr)
```